framework/web/UrlRule.php

```diff
@@ -256,6 +355,25 @@ public function parseRequest($manager, $request)
         }
         $matches = $this->substitutePlaceholderNames($matches);
 
+        // if routing succeeded (some rule matched), check whether pathInfo
+        //  was changed during canonization. If so, take the required action.
+        if ($pathInfoCanonized != $pathInfoOrig) {
+            switch ($this->canonize['action']) {
+                case 'redirect':
+                    $url = '/' . $pathInfoCanonized
+                        . (Yii::$app->request->queryString ? '?' . Yii::$app->request->queryString : '');
+                    die('redirect ' . $url);
+                    Yii::$app->getResponse()->redirect($url)->send();
+                    break;
```

Comment:
That sounds not much better than current solution in terms of developer actions required: https://github.com/samdark/yii2-cookbook/blob/master/book/handling-trailing-slash-in-urls.md